integration/client-data-test.ts

```diff
@@ -0,0 +1,446 @@
+      expect(html).toMatch("Child Server Loader");
+    });
+
+    test("parent.clientLoader/child.clientLoader.hydrate", async ({ page }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({
+            parentClientLoader: true,
+            parentClientLoaderHydrate: false,
+            childClientLoader: true,
+            childClientLoaderHydrate: true,
+          }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Renders child fallback on initial render and calls child clientLoader
+      // Does not call parent clientLoader due to lack of HydrateFallback
+      await app.goto("/parent/child");
+      let html = await app.getHtml("main");
+      expect(html).toMatch("Parent Server Loader");
+      expect(html).toMatch("Child Fallback");
+      expect(html).not.toMatch("Child Server Loader");
+
+      await page.waitForSelector("#child-data");
+      html = await app.getHtml("main");
+      expect(html).not.toMatch("Child Fallback");
+      expect(html).toMatch("Parent Server Loader");
+      expect(html).toMatch("Child Server Loader (mutated by client)");
+    });
+
+    test("parent.clientLoader.hydrate/child.clientLoader.hydrate", async ({
+      page,
+    }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({
+            parentClientLoader: true,
+            parentClientLoaderHydrate: true,
+            childClientLoader: true,
+            childClientLoaderHydrate: true,
+          }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Renders parent fallback on initial render and calls both clientLoader's
+      await app.goto("/parent/child");
+      let html = await app.getHtml("main");
+      expect(html).toMatch("Parent Fallback");
+      expect(html).not.toMatch("Parent Server Loader");
+      expect(html).not.toMatch("Child Fallback");
+      expect(html).not.toMatch("Child Server Loader");
+
+      await page.waitForSelector("#child-data");
+      html = await app.getHtml("main");
+      expect(html).not.toMatch("Parent Fallback");
+      expect(html).not.toMatch("Child Fallback");
+      expect(html).toMatch("Parent Server Loader (mutated by client)");
+      expect(html).toMatch("Child Server Loader (mutated by client)");
+    });
+
+    test("handles synchronous client loaders", async ({ page }) => {
+      let fixture = await createFixture({
+        files: getFiles({
+          parentClientLoader: false,
+          parentClientLoaderHydrate: false,
+          childClientLoader: false,
+          childClientLoaderHydrate: false,
+          parentAdditions: js`
+            export function clientLoader() {
+              return { message: "Parent Client Loader" };
+            }
+            clientLoader.hydrate=true
+            export function HydrateFallback() {
+              return <p>Parent Fallback</p>
+            }
```

Comment:
Is the fallback going to be called `HydrateFallback`? Also we will need to add `clientLoader.hydrate = true` to every clientLoader we want to run on hydration?

Reply:
> Is the fallback going to be called HydrateFallback

Yeah, we went with `HydrateFallback` for a few reasons:
* It's only relevant for initial document hydration when you opt-into the `clientLoader` executing on hydration (via `clientLoader.hydrate=true`)
* `Fallback`/`ClientFallback` are misleading because they might imply that it will render during subsequent SPA navs which is not the case
* The verbiage aligns with `clientLoader.hydrate` and `ReactDOM.hydrateRoot`

> Also we will need to add clientLoader.hydrate = true to every clientLoader we want to run on hydration?

Yeah:
* The default behavior matches the current remix behavior.  Every route has a client loader in Remix today which is just the `fetch()` call - and those _do not_ run on hydration because we already have the data.  Exporting a `clientLoader` is basically you taking more control over this existing client side loader - but it does not change the behavior.
* We believe `clientLoader` as an optimization is the more frequent use case (basically where clientLoader doesn't change the shape of the returned data)
  * Client side caching
  * Hitting an API directly on a SPA nav and bypassing the Remix BFF network hop

We think the use case where you want to run on hydration is likely less frequent since it eliminates your ability to SSR content, introduces more potential for CLS, etc.